Add RSS feed [#932]

genehack opened the pull request:
## Description of proposed changes

Add RSS generation script and hook it into build [#932]

Note: Atom and JSON feeds will be added to this once any review of the overall metadata is complete.

* Add function to `blogPosts.js` to read blog post and generate feed with the 10 most recent posts into the `public` directory
* Add new top-level script that calls the new `generateBlogFeeds` function
* Add package.json:scripts target to run this script
* Hook that package.json:scripts target into build.sh
* Ignore the new build-time generated file

## Related issue(s)

Closes #932 

## Checklist

<!--
Make sure checks are successful at the bottom of the PR.

If applicable, add:
- any changes to existing tests
- any additional manual testing to confirm changes

Please add a note if you need help with adding tests.
-->

- [x] Checks pass
- [x] Check if changes affect the [resource index JSON revision](https://docs.nextstrain.org/projects/nextstrain-dot-org/en/latest/resource-collection.html#resource-index-revisions)

<!-- 🙌 Thank you for contributing to Nextstrain! ✨ -->

genehack commented:
This is my first time mucking with the build/infra side of nextstrain.org, so I'm marking this as draft and tagging @jameshadfield and @victorlin for any feedback.

There are a couple additional (fairly trivial) changes needed to get Atom and JSON feeds working, which I will probably land tomorrow, but I wanted to get this pushed so folks could give me initial feedback. There are a couple of FIXMEs noted, and I'm open to any and all feedback about what's being included in the feed.

jameshadfield commented on static-site/src/util/blogPosts.js:
```diff
@@ -2,7 +2,9 @@ import fs from 'fs';
-import { startCase } from "lodash"
+import lodash from 'lodash';
+
+const { startCase } = lodash;
```
`import { startCase } from "lodash"` often allows tree-shaking (cf. `import lodash from 'lodash';`). I can't remember the details of what nextjs is doing here / what we've asked it to do. Might be worth [analysing the bundle](https://nextjs.org/docs/pages/building-your-application/optimizing/package-bundling#analyzing-javascript-bundles) here to check.

genehack replied:
> `import { startCase } from "lodash"` often allows tree-shaking (cf. `import lodash from 'lodash';`). I can't remember the details of what nextjs is doing here / what we've asked it to do. 

When this file contains `import { startCase } from 'lodash';`, running the feed generation script produces this error:

```javascript
> npm run build:feeds                                                                                                                 v20.9.0

> build:feeds
> node scripts/generate-blog-feeds.js

file:///Users/genehack/nextstrain/nextstrain.org/static-site/src/util/blogPosts.js:9
import { startCase } from 'lodash';
         ^^^^^^^^^
SyntaxError: Named export 'startCase' not found. The requested module 'lodash' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'lodash';
const { startCase } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:131:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:213:5)
    at async ModuleLoader.import (node:internal/modules/esm/loader:316:24)
    at async loadESM (node:internal/process/esm_loader:34:7)
    at async handleMainPromise (node:internal/modules/run_main:66:12)

Node.js v20.9.0
```

(Trying to evaluate the `blogPosts.js` file directly, in the `static-site` directory, generates the exact same error.)

I'm not sure why this error doesn't happen when the file is included in the 2 JSX files that consume it. 

> Might be worth [analysing the bundle](https://nextjs.org/docs/pages/building-your-application/optimizing/package-bundling#analyzing-javascript-bundles) here to check.

The only place this file gets included in the site is in 2 JSX files that are used to produce static site files — I don't think the `lodash` Javascript would need to end up in the bundle? But maybe I'm confused about how the "static" file generation works in the production build.

jameshadfield replied:
> The only place this file gets included in the site is in 2 JSX files that are used to produce static site files — I don't think the lodash Javascript would need to end up in the bundle? But maybe I'm confused about how the "static" file generation works in the production build.

In this case it's not ending up in the bundle - next.js is (correctly) not including it even tho the files which import it themselves export frontend JSX.

> Might be worth [analysing the bundle](https://nextjs.org/docs/pages/building-your-application/optimizing/package-bundling#analyzing-javascript-bundles) here to check.

I did this to double check and it identified some other issues. Will make an issue.

jameshadfield replied:
> I did this to double check and it identified some other issues. Will make an issue.

Cancel that, was looking at stat sizes not parsed/compressed sizes.

jameshadfield reviewed (commented):
`npm run build:feeds` ran without trouble for me locally. I don't have experience with RSS spec, so no comments from me on that. How are feeds normally discovered? I.e. should we not add an icon on the blog somewhere linking to the (newly added) `/feed.xml` URL? Finally my presumption of the URL would be `/blog/feed`.

victorlin commented on static-site/src/util/blogPosts.js:
```diff
@@ -36,6 +39,59 @@ export function getBlogPosts() {
+
+  if (posts && posts.length > 0) {
+    for (const post of posts) {
+      if (post) { // make the type checker happier
+        const url = `${siteURL}/${post.blogUrlName}`;
```
Did you see linting errors? I removed the conditionals and `npm run lint:static-site` still runs fine.

```js
for (const post of posts) {
const url = `${siteURL}/${post.blogUrlName}`;

feed.addItem({
    title: post.title,
    id: url,
    link: url,
    author: post.author,
    date: new Date(post.date),
});
}
```

genehack replied:
I didn't see _linting_ errors, but my editor typechecks javascript code too. 

Reviewing it in the cold light of morning, the `if (posts && posts.length > 0)` guard isn't needed (and I'll remove it). 

The interior `if (post)` guard _is_ needed, because [elements in the returned list _might_ be `false`](https://github.com/nextstrain/nextstrain.org/blob/master/static-site/src/util/blogPosts.js#L27) instead of an object, which would cause the element accesses in `feed.addItem()` to fail.

genehack commented:
> `npm run build:feeds` ran without trouble for me locally. I don't have experience with RSS spec, so no comments from me on that.

W3C has a [feed validator](https://validator.w3.org/feed/#validate_by_input), that you can copypasta a file into. Once I add the Atom and JSON feed versions, I will validate each of them. (A [JSON Feed validator](https://validator.jsonfeed.org/) also exists.)

> How are feeds normally discovered? I.e. should we not add an icon on the blog somewhere linking to the (newly added) `/feed.xml` URL?

We could add an icon (maybe next to the `BLOG` title in the sidebar?), but generally, feed autodiscovery works via including something like this in the `<head>` of (at least) the index page -- but I would advocate for adding it to `https://nexstrain.org` as well as all URLs that start with `https://nextstrain.org/blog`:

```html
<link href="https://genehack.blog/feed.rss2" rel="alternate" title="RSS2 feed for genehack.blog" type="application/rss+xml">
<link href="https://genehack.blog/atom.xml" rel="alternate" title="Atom feed for genehack.blog  type="application/atom+xml">
<link href="https://genehack.blog/feed.json" rel="alternate" title="JSON feed for genehack.blog" type="application/json">"
```

(Examples from my personal blog just because I had them handy.)

> Finally my presumption of the URL would be `/blog/feed`.

Containing it under the `/blog` stem is fine, but note there will be multiple feed versions, so we'll have `/blog/feed/rss2.xml`, `/blog/feed/atom.xml`, etc.

genehack commented:
Okay, I have added Atom and JSON feeds, linked them into the `<head>` of the template, and updated various other bits.

The one thing remaining is to add a top-level "description" to the feed. This _could_ be "Real-time tracking of pathogen evolution", as we use as the top-level description, but I'm not sure that's appropriate for the blog? 

Suggestions? (/cc @trvrb for visibility on the description question.)

genehack commented:
> W3C has a [feed validator](https://validator.w3.org/feed/#validate_by_input), that you can copypasta a file into. Once I add the Atom and JSON feed versions, I will validate each of them. 

The Atom and RSS feeds validate.

> (A [JSON Feed validator](https://validator.jsonfeed.org/) also exists.)

The JSON feed needs a little work. 

I think I'm also going to add the full text content of the posts to the feeds as well, unless somebody thinks that's too much.

genehack commented:
> The JSON feed needs a little work.

The JSON feed validates now. 

> I think I'm also going to add the full text content of the posts to the feeds as well, unless somebody thinks that's too much.

And the content chunks have been added. 

This is ready to go, I think...

victorlin commented on static-site/src/components/layout.jsx:
```diff
@@ -23,6 +23,9 @@ export default class MainLayout extends React.Component {
           <title>{`${siteTitle}`}</title>
           <meta name="description" content={siteDescription} />
           <link rel="me" href="https://mstdn.science/@nextstrain" />
+          <link href={`${siteUrl}/blog/atom.xml`} rel="alternate" title="Atom feed for nextstrain.org/blog" type="application/atom+xml" />
+          <link href={`${siteUrl}/blog/feed.json`} rel="alternate" title="JSON feed for nextstrain.org/blog" type="application/json" />
+          <link href={`${siteUrl}/blog/rss2.xml`} rel="alternate" title="RSS2 feed for nextstrain.org/blog" type="application/rss+xml" />
```
_non-blocking_

All these (including mastodon link) should be conditioned on SiteConfig's `groupsApp` being `false`, but no need to block this PR on that.

genehack replied:
done.

victorlin commented on static-site/src/util/blogPosts.js:
```diff
@@ -36,6 +42,76 @@ export function getBlogPosts() {
+  const siteURL = "https://nextstrain.org/blog";
+
+  const feed = new Feed({
+    title: "Nextstrain.org Blog",
```
_non-blocking_

```suggestion
    title: "Nextstrain Blog",
```

This is the blog of Nextstrain that happens to be on nextstrain.org, not a blog specifically about nextstrain.org

victorlin commented on static-site/src/util/blogPosts.js:
```diff
@@ -36,6 +42,76 @@ export function getBlogPosts() {
+
+  const feed = new Feed({
+    title: "Nextstrain.org Blog",
+    description: "FIXME",
```
```suggestion
    description: "Updates from the Nextstrain core team",
```

Doesn't have to be this, just something other than `FIXME` 😄

genehack replied:
I've been waiting for somebody to suggest something! :grin:

victorlin commented on static-site/src/util/blogPosts.js:
```diff
@@ -36,6 +42,76 @@ export function getBlogPosts() {
+    description: "FIXME",
+    id: siteURL,
+    link: siteURL,
+    image: "https://nextstrain.org/_next/static/media/nextstrain-logo-small.dcf6bdd2.png",
```
```suggestion
    image: "https://nextstrain.org/nextstrain-logo-small.png",
```

Use `static-site/public/nextstrain-logo-small.png` which is served at the top level using the filename as-is, not `static-site/static/logos/nextstrain-logo-small.png` which is bundled by webpack with a hash in the filename that may change over time.

victorlin commented on static-site/src/util/blogPosts.js:
```diff
@@ -36,6 +42,76 @@ export function getBlogPosts() {
+ */
+export async function generateBlogFeeds() {
+  const feedDir = `${__dirname}/../../public/blog`;
+  const siteURL = "https://nextstrain.org/blog";
```
```suggestion
  const blogURL = "https://nextstrain.org/blog";
```

I was confused why these paths seemed to look different:

https://github.com/nextstrain/nextstrain.org/blob/5709c3ce11f60cf66fa4ae3604643018e70e2e00/static-site/src/util/blogPosts.js#L62

https://github.com/nextstrain/nextstrain.org/blob/5709c3ce11f60cf66fa4ae3604643018e70e2e00/static-site/src/components/layout.jsx#L26

then I realized that `siteURL` and `siteUrl` are not the same. Renaming `siteURL` (or even better, [putting the Atom/JSON/RSS2 paths in shared variables](https://github.com/nextstrain/nextstrain.org/pull/990#discussion_r1731861727)) would reduce confusion.

genehack replied:
Kinda ended up doing both of these things...

victorlin commented on static-site/src/components/layout.jsx:
```diff
@@ -23,6 +23,9 @@ export default class MainLayout extends React.Component {
           <title>{`${siteTitle}`}</title>
           <meta name="description" content={siteDescription} />
           <link rel="me" href="https://mstdn.science/@nextstrain" />
+          <link href={`${siteUrl}/blog/atom.xml`} rel="alternate" title="Atom feed for nextstrain.org/blog" type="application/atom+xml" />
+          <link href={`${siteUrl}/blog/feed.json`} rel="alternate" title="JSON feed for nextstrain.org/blog" type="application/json" />
+          <link href={`${siteUrl}/blog/rss2.xml`} rel="alternate" title="RSS2 feed for nextstrain.org/blog" type="application/rss+xml" />
         </Helmet>
```
_non-blocking_

These href values could be more tighly coupled with where they are defined:

https://github.com/nextstrain/nextstrain.org/blob/5709c3ce11f60cf66fa4ae3604643018e70e2e00/static-site/src/util/blogPosts.js#L61-L65

They could be top-level constants in `blogPosts.js` and imported here.

victorlin commented on package.json:
```diff
@@ -47,6 +48,7 @@
     "express-naked-redirect": "^0.1.2",
     "express-session": "^1.16.2",
     "express-static-gzip": "^0.2.2",
+    "feed": "^4.2.2",
```
_non-blocking_

I wondered if `feed` should be under `devDependencies` because the feed is only updated during build time, but realized that's not possible with the way this has been implemented. If `scripts/generate-blog-feeds.js` were the only thing that used `blogPosts.js` then it should be a dev dependency. However, Next.js uses `getBlogPosts()` (in the same file) during runtime.

If this is reorganized so that `generateBlogFeeds()` is not in the same file as `getBlogPosts()`, `feed` could be made a dev dependency. I'll let you decide if that's worthwhile.

genehack commented:
Okay, @victorlin, refactored so that `feed` can be a devDep, and adjusted various constants to be imported from SiteConfig (which turned into maybe a larger than expected refactoring due to the need to access things from a non-JSX file...)

I _think_ this is overall cleaner now, but also maybe a little more convoluted? Interested to hear what you think.

victorlin commented on static-site/data/BaseConfig.js:
```diff
@@ -0,0 +1,31 @@
+
+export const groupsApp = false;
+
+export const groupsTitle = "Scalable Sharing with Nextstrain Groups";
+
```
I believe CDC's internal "groups app instance" has customized these in SiteConfig.jsx. This would be a breaking change for them, which should be fine as long as we give them a heads up on the `#nextstrain-cdc` Slack channel.

genehack replied:
Tracked in #1001

victorlin commented on static-site/src/util/blogFeeds.js:
```diff
@@ -0,0 +1,76 @@
+    }
+  }
+
+  const feeds = [
+    {
+      file: `${feedDir}/atom.xml`,
+      data: feed.atom1,
+    },
+    {
+      file: `${feedDir}/feed.json`,
+      data: feed.json1,
+    },
+    {
+      file: `${feedDir}/rss2.xml`,
+      data: feed.rss2,
+    },
+  ];
```
Get the `file` values from BaseConfig?

https://github.com/nextstrain/nextstrain.org/blob/43a922fb76960f818e28cb64a0b0139c27ec81ff/static-site/data/BaseConfig.js#L27-L31

genehack replied:
refactored.

victorlin reviewed (commented):
Quickly skimmed before heading off to lunch

genehack commented:
> Quickly skimmed before heading off to lunch

* stripped comment
* created [tracking issue](#1001) for CDC notification about config rearrangement
* refactored blog feed data structure to enable SSoT design
* updated groupsApp.md docs to reflect SiteConfig/BaseConfig split

victorlin commented on static-site/data/BaseConfig.js:
Re: https://github.com/nextstrain/nextstrain.org/pull/990#issuecomment-2313359755

> I _think_ this is overall cleaner now, but also maybe a little more convoluted? Interested to hear what you think.

Separating dev vs. runtime deps seems more important than keeping all config in SiteConfig.jsx, so this seems fine to me.

victorlin reviewed (approved):
Code looks good. I'm not familiar enough with feeds so I don't know how to properly test the Atom/JSON/RSS files, but I trust your judgement there!

genehack commented:
> I'm not familiar enough with feeds so I don't know how to properly test the Atom/JSON/RSS files, but I trust your judgement there!

They were validating in the preview build in the W3C and JSON feed validators. I'll doublecheck in the morning before I merge.
